core/src/main/scala-3/caliban/schema/SchemaDerivation.scala

```diff
@@ -171,26 +171,22 @@ trait CommonSchemaDerivation {
       else if (isInput) mkInputObject[R](annotations, fields, info)(isInput, isSubscription)
       else mkObject[R](annotations, fields, info)(isInput, isSubscription)
 
-    override private[schema] lazy val resolveFieldLazily: Boolean = (!isValueType) && fields.isEmpty
-
     def resolve(value: A): Step[R] =
       if (fields.isEmpty) PureStep(EnumValue(name))
       else if (isValueType) resolveValueType(value)
-      else resolveObject(value)
+      else MetadataFunctionStep[R](f => resolveObject(value, f.fieldNames))
```

Comment:
I ended up moving this here for 2 reasons:

1. It's slightly more performant since `fieldNames` is a lazy val and there is a small overhead when evaluating if a lazy val was initialized. This can add up when it's evaluated for each field.
2. It makes profiling easier to follow since the logic in `resolveObject` won't be wrapped within an anonymous function